server/src/main/java/com/flipkart/varadhi/services/DlqService.java

```
@@ -34,37 +42,55 @@ public CompletableFuture<SubscriptionOperation> unsideline(
-                            ConsumerApi consumer = consumerFactory.getInstance(assignment.getConsumerId());
-                            return consumer.getMessages(messagesRequest).whenComplete((messages, t) -> {
-                                // write partial response i.e. return messages from single shard or error.
-                                GetMessagesResponse shardResponse = GetMessagesResponse.of(assignment.getShardId());
-                                if (t == null) {
-                                    shardResponse.setMessages(messages);
-                                } else {
-                                    log.error(
-                                            "Failed to getMessages for {}(shard:{}). Error:{}", subscription.getName(),
-                                            assignment.getShardId(), t.getMessage()
-                                    );
-                                    shardResponse.setError(t.getMessage());
-                                }
-                                log.info("Sending a GetMessage response for {}", assignment);
-                                responseWriter.accept(shardResponse);
-                            });
-                        }).toArray(CompletableFuture[]::new)));
+        ShardDlqMsgResponseCollector finalResponse = new ShardDlqMsgResponseCollector();
+        boolean isRequestByTimeStamp = earliestFailedAt != UNSPECIFIED_TS;
+        return controllerClient.getShardAssignments(subscription.getName()).thenCompose(assignments -> {
+            List<CompletableFuture<ShardDlqMessageResponse>> shardFutures = new ArrayList<>();
+            for (Assignment a : assignments.getAssignments()) {
+                int shardId = a.getShardId();
+                if (!isRequestByTimeStamp && !pageMarkers.hasMarker(shardId)) {
+                    log.info("Shard {} has no markers, skipping getMessages().", shardId);
+                    continue;
+                }
```

We are skipping a shard if info not provided. Is this intentional?

Should we just fail the request outright as bad request if a marker is missing for any shard?

---

I would guess,  this may be helpful in fetching messages from a partial set of shards.

---

yes. If marker is not available, it is assumed there are no further messages to be returned and hence such shards are skipped. It is also assumed users are not manipulating the markers as they are abstracted from them.